File: frontend/lib/models/imovel.dart
import 'dart:convert';

class AnaliseIA {
  final String? resumoExecutivo;
  final List<String> pontosPositivos;
  final List<String> pontosAtencao;
  final List<String> riscos;
  final String? recomendacaoFinal;
  final String? classificacao;

  AnaliseIA({
    this.resumoExecutivo,
    this.pontosPositivos = const [],
    this.pontosAtencao = const [],
    this.riscos = const [],
    this.recomendacaoFinal,
    this.classificacao,
  });

  factory AnaliseIA.fromJson(Map<String, dynamic> json) {
    List<String> parseList(dynamic val) {
      if (val == null) return [];
      if (val is List) return val.cast<String>();
      try {
        return (jsonDecode(val as String) as List).cast<String>();
      } catch (_) {
        return [];
      }
    }

    return AnaliseIA(
      resumoExecutivo: json['resumo_executivo'],
      pontosPositivos: parseList(json['pontos_positivos']),
      pontosAtencao: parseList(json['pontos_atencao']),
      riscos: parseList(json['riscos_juridicos']),
      recomendacaoFinal: json['recomendacao_final'],
      classificacao: json['classificacao'],
    );
  }
}

class Imovel {
  final int id;
  final String endereco;
  final String? bairro;
  final String cidade;
  final String estado;
  final String tipo;
  final double? areaConstuida;
  final double? areaTereno;
  final bool ocupado;
  final double valorAvaliacao;
  final double valorLance;
  final double? valorMercado;
  final double score;
  final double? desconto;
  final double? lucroPotencial;
  final double? roi;
  final int? paybackMeses;
  final DateTime? dataLeilao;
  final String? urlImovel;
  final String? urlEdital;
  final bool noRadar;
  final bool isFavorito;
  final double? latitude;
  final double? longitude;
  final AnaliseIA? analiseIA;
  final String fonte;

  Imovel({
    required this.id,
    required this.endereco,
    this.bairro,
    required this.cidade,
    required this.estado,
    required this.tipo,
    this.areaConstuida,
    this.areaTereno,
    required this.ocupado,
    required this.valorAvaliacao,
    required this.valorLance,
    this.valorMercado,
    required this.score,
    this.desconto,
    this.lucroPotencial,
    this.roi,
    this.paybackMeses,
    this.dataLeilao,
    this.urlImovel,
    this.urlEdital,
    required this.noRadar,
    required this.isFavorito,
    this.latitude,
    this.longitude,
    this.analiseIA,
    required this.fonte,
  });

  factory Imovel.fromJson(Map<String, dynamic> json) => Imovel(
        id: json['id'],
        endereco: json['endereco'] ?? '',
        bairro: json['bairro'],
        cidade: json['cidade'] ?? 'Aracaju',
        estado: json['estado'] ?? 'SE',
        tipo: json['tipo'] ?? 'outro',
        areaConstuida: (json['area_construida'] as num?)?.toDouble(),
        areaTereno: (json['area_terreno'] as num?)?.toDouble(),
        ocupado: json['ocupado'] ?? false,
        valorAvaliacao: (json['valor_avaliacao'] as num).toDouble(),
        valorLance: (json['valor_lance_minimo'] as num).toDouble(),
        valorMercado: (json['valor_mercado_estimado'] as num?)?.toDouble(),
        score: (json['score'] as num?)?.toDouble() ?? 0,
        desconto: (json['desconto_percentual'] as num?)?.toDouble(),
        lucroPotencial: (json['lucro_potencial'] as num?)?.toDouble(),
        roi: (json['roi_estimado'] as num?)?.toDouble(),
        paybackMeses: json['payback_meses'],
        dataLeilao: json['data_leilao'] != null ? DateTime.tryParse(json['data_leilao']) : null,
        urlImovel: json['url_imovel'],
        urlEdital: json['url_edital'],
        noRadar: json['no_radar'] ?? false,
        isFavorito: json['is_favorito'] ?? false,
        latitude: (json['latitude'] as num?)?.toDouble(),
        longitude: (json['longitude'] as num?)?.toDouble(),
        analiseIA: json['analise_ia'] != null ? AnaliseIA.fromJson(json['analise_ia']) : null,
        fonte: json['fonte'] ?? 'outro',
      );

  String get classificacaoLabel {
    switch (analiseIA?.classificacao) {
      case 'excelente': return 'Excelente Oportunidade';
      case 'boa': return 'Boa Oportunidade';
      case 'moderada': return 'Oportunidade Moderada';
      case 'alto_risco': return 'Alto Risco';
      default: return 'Em análise';
    }
  }
}

class ImovelListResponse {
  final int total;
  final int page;
  final int perPage;
  final int pages;
  final List<Imovel> items;

  ImovelListResponse({
    required this.total,
    required this.page,
    required this.perPage,
    required this.pages,
    required this.items,
  });

  factory ImovelListResponse.fromJson(Map<String, dynamic> json) => ImovelListResponse(
        total: json['total'],
        page: json['page'],
        perPage: json['per_page'],
        pages: json['pages'],
        items: (json['items'] as List).map((i) => Imovel.fromJson(i)).toList(),
      );
}

File: backend/app/schemas/imovel.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from app.models.imovel import TipoImovelEnum, FonteEnum


class AnaliseIAOut(BaseModel):
    resumo_executivo: Optional[str] = None
    pontos_positivos: Optional[str] = None
    pontos_atencao: Optional[str] = None
    riscos_juridicos: Optional[str] = None
    recomendacao_final: Optional[str] = None
    classificacao: Optional[str] = None

    class Config:
        from_attributes = True


class ImovelBase(BaseModel):
    endereco: str
    bairro: Optional[str] = None
    cidade: str = "Aracaju"
    estado: str = "SE"
    cep: Optional[str] = None
    tipo: TipoImovelEnum
    area_construida: Optional[float] = None
    area_terreno: Optional[float] = None
    quartos: Optional[int] = None
    banheiros: Optional[int] = None
    vagas: Optional[int] = None
    ocupado: bool = False
    valor_avaliacao: float
    valor_lance_minimo: float
    valor_divida: Optional[float] = None
    data_leilao: Optional[datetime] = None
    data_segunda_praca: Optional[datetime] = None
    url_edital: Optional[str] = None
    url_imovel: Optional[str] = None
    fotos: Optional[str] = None
    numero_edital: Optional[str] = None
    numero_processo: Optional[str] = None
    matricula: Optional[str] = None
    fonte: FonteEnum = FonteEnum.outro
    leiloeiro: Optional[str] = None


class ImovelCreate(ImovelBase):
    pass


class ImovelOut(ImovelBase):
    id: int
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    score: float
    desconto_percentual: Optional[float] = None
    lucro_potencial: Optional[float] = None
    roi_estimado: Optional[float] = None
    payback_meses: Optional[int] = None
    vgv_m2: Optional[float] = None
    valor_mercado_estimado: Optional[float] = None
    no_radar: bool
    created_at: datetime
    analise_ia: Optional[AnaliseIAOut] = None
    is_favorito: Optional[bool] = False

    class Config:
        from_attributes = True


class ImovelFiltros(BaseModel):
    bairro: Optional[str] = None
    cidade: Optional[str] = None
    estado: Optional[str] = None
    tipo: Optional[TipoImovelEnum] = None
    valor_minimo: Optional[float] = None
    valor_maximo: Optional[float] = None
    area_minima: Optional[float] = None
    area_maxima: Optional[float] = None
    score_minimo: Optional[float] = None
    ocupado: Optional[bool] = None
    no_radar: Optional[bool] = None
    fonte: Optional[FonteEnum] = None
    page: int = 1
    per_page: int = 20


class ImovelListResponse(BaseModel):
    total: int
    page: int
    per_page: int
    pages: int
    items: List[ImovelOut]
